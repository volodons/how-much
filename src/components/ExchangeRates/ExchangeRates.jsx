import { Box, Grid, Typography, styled } from "@mui/material";
import Flag from "react-world-flags";
import StarOutlineIcon from "@mui/icons-material/StarOutline";
import "typeface-lobster";

const StyledBox = styled(Box)`
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 2rem;
    padding: 2rem;
    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    border-radius: 10px;
    background: linear-gradient(to right, #0390fc, #562fff);
    transition: box-shadow 0.3s ease-in-out;
    &:hover {
        transform: scale(1.02);
    }
    transition: box-shadow 0.3s ease-in-out;
    &:hover {
        box-shadow: 4px 4px 8px rgba(0, 0, 0, 0.5);
    }
`;

const StyledGrid = styled(Grid)`
    justify-content: space-between;
    align-items: center;
`;

const StyledTypographyHeading = styled(Typography)`
    margin-bottom: 1rem;
    text-align: center;
    color: #f5f7ff;
    text-decoration: underline;
    font-family: Lobster, sans-serif;
    text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
    }
`;

const StyledTypographyText = styled(Typography)`
    font-size: 3rem;
    color: #f5f7ff;
    text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
    }
    transition: color 0.3s ease-in-out;
    &:hover {
        opacity: 0.8;
    }
`;

const StyledIcon = styled(StarOutlineIcon)`
    width: 3rem;
    height: 3rem;
    color: #ffd700;
    &:hover {
        opacity: 0.8;
    }
`;

const StyledFlag = styled(Flag)`
    width: 6rem;
    border-radius: 5px;
    transition: color 0.3s ease-in-out;
    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.3);
    &:hover {
        box-shadow: 3px 3px 6px rgba(0, 0, 0, 0.4);
        opacity: 0.8;
    }
`;

const ExchangeRates = () => {
    return (
        <StyledBox>
            <StyledTypographyHeading variant="h2">
                Exchange Rates
            </StyledTypographyHeading>
            <Grid container>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="USA" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>5</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="EU" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>10</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="JPN" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>15</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="GBR" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>20</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="CHN" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>25</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="AUS" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>30</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="CAN" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>35</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="CHE" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>40</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="POL" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>45</StyledTypographyText>
                    </Grid>
                </StyledGrid>
                <StyledGrid container>
                    <Grid item>
                        <StyledIcon />
                    </Grid>
                    <Grid item>
                        <StyledFlag code="ISR" />
                    </Grid>
                    <Grid item>
                        <StyledTypographyText>50</StyledTypographyText>
                    </Grid>
                </StyledGrid>
            </Grid>
        </StyledBox>
    );
};

export default ExchangeRates;
