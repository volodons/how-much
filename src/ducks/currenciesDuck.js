import { createSlice } from '@reduxjs/toolkit';

import ALL_CURRENCIES from '../const/currencies';

const initialState = {
    allCurrencies: [...ALL_CURRENCIES],
    featuredCurrencies: [],
};

const currenciesSlice = createSlice({
    name: 'Currencies',
    initialState,
    reducers: {
        ADD_TO_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = [
                ...state.featuredCurrencies,
                action.payload,
            ];
            state.allCurrencies = state.allCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
        },
        REMOVE_FROM_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = state.featuredCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
            state.allCurrencies = [...state.allCurrencies, action.payload];
        },
    },
});

const { actions, reducer } = currenciesSlice;

export const currenciesActions = actions;
export const currenciesReducer = reducer;
