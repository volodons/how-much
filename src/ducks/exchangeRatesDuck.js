import { createSlice } from "@reduxjs/toolkit";

const exchangeRatesSlice = createSlice({
    name: "Exchange Rates",
    initialState: {
        amountBeforeConversion: null,
        amountAfterConversion: null,
        baseCurrency: null,
        targetCurrency: null,
        exchangeRate: null,
    },
    reducers: {
        SET_AMOUNT_BEFORE_CONVERSION: (state, action) => {
            state.amountBeforeConversion = action.payload;
        },
        SET_AMOUNT_AFTER_CONVERSION: (state, action) => {
            state.amountAfterConversion = action.payload;
        },
        SET_BASE_CURRENCY: (state, action) => {
            state.baseCurrency = action.payload;
        },
        SET_TARGET_CURRENCY: (state, action) => {
            state.targetCurrency = action.payload;
        },
        SET_EXCHANGE_RATE: (state, action) => {
            state.exchangeRate = action.payload;
        },
        FETCH_EXCHANGE_RATES: (state) => state,
    },
});

const { actions, reducer } = exchangeRatesSlice;

export { actions as exchangeRatesSlice, reducer as exchangeRatesReducer };
