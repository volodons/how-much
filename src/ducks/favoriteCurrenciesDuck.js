import { createSlice } from "@reduxjs/toolkit";

const favoriteCurrenciesSlice = createSlice({
    name: "Favorite Currencies",
    initialState: {
        favoriteCurrencies: [],
    },
    reducers: {
        ADD_TO_FAVORITES: (state, action) => {
            state.favoriteCurrencies = [
                ...state.favoriteCurrencies,
                action.payload,
            ];
        },
        REMOVE_FROM_FAVORITES: (state, action) => {
            state.favoriteCurrencies = state.favoriteCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
        },
    },
});

const { actions, reducer } = favoriteCurrenciesSlice;

export {
    actions as favoriteCurrenciesActions,
    reducer as favoriteCurrenciesReducer,
};
