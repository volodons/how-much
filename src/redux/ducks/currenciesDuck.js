import { createSlice } from '@reduxjs/toolkit';
import { call, put, takeLatest } from 'redux-saga/effects';

import { fetchAllCurrencies } from '../api';

const initialState = {
    allCurrencies: [],
    featuredCurrencies: [],
    error: null,
};

const currenciesSlice = createSlice({
    name: 'Currencies',
    initialState,
    reducers: {
        FETCH_ALL_CURRENCIES: (state) => state,
        FETCH_ALL_CURRENCIES_FAILURE: (state, action) => {
            state.error = action.payload;
        },
        SET_ALL_CURRENCIES: (state, action) => {
            state.allCurrencies = [...action.payload];
        },
        ADD_TO_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = [
                ...state.featuredCurrencies,
                action.payload,
            ];
            state.allCurrencies = state.allCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
        },
        REMOVE_FROM_FEATURED_CURRENCIES: (state, action) => {
            state.featuredCurrencies = state.featuredCurrencies.filter(
                (currency) => currency.id !== action.payload.id
            );
            state.allCurrencies = [...state.allCurrencies, action.payload];
        },
    },
});

function* fetchAllCurrenciesSaga() {
    try {
        const response = yield call(fetchAllCurrencies);
        yield put(currenciesActions.SET_ALL_CURRENCIES(response));
    } catch (error) {
        yield put(
            currenciesActions.FETCH_ALL_CURRENCIES_FAILURE(error.message)
        );
    }
}

export function* watchAllCurrencies() {
    yield takeLatest(
        currenciesActions.FETCH_ALL_CURRENCIES,
        fetchAllCurrenciesSaga
    );
}

const { actions, reducer } = currenciesSlice;

export const currenciesActions = actions;
export const currenciesReducer = reducer;
