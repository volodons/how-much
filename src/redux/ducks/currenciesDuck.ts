import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { call, put, takeLatest } from 'redux-saga/effects';

import {
    fetchExchangeRatesCurrencies,
    fetchExchangeRates,
    fetchBaseCurrency,
} from '../../api';

interface Currency {
    id: number;
    code: string;
    exchangeRate: number;
    featured: boolean;
}

interface ExchangeRatesCurrenciesResponse {
    currenciesWithData: Currency[];
    filteredCurrencyCodes: string[];
}

interface ExchangeRate {
    [key: string]: {
        code: string;
        value: number;
    };
}

interface CurrenciesState {
    currencies: Currency[];
    currencyCodes: string[];
    baseCurrency: string;
    error: string | null;
}

const initialState: CurrenciesState = {
    currencies: [],
    currencyCodes: [],
    baseCurrency: '',
    error: null,
};

const currenciesSlice = createSlice({
    name: 'Currencies',
    initialState,
    reducers: {
        FETCH_CURRENCIES: (state) => state,
        FETCH_CURRENCIES_FAILURE: (state, action: PayloadAction<string>) => {
            state.error = action.payload;
        },
        SET_CURRENCIES: (
            state,
            action: PayloadAction<ExchangeRatesCurrenciesResponse>
        ) => {
            state.currencies = [...action.payload.currenciesWithData];
            state.currencyCodes = [...action.payload.filteredCurrencyCodes];
        },
        FETCH_BASE_CURRENCY: (state) => state,
        FETCH_BASE_CURRENCY_FAILURE: (state, action: PayloadAction<string>) => {
            state.error = action.payload;
        },
        SET_BASE_CURRENCY: (state, action: PayloadAction<string>) => {
            state.baseCurrency = action.payload;
        },
        FETCH_EXCHANGE_RATES: (state) => state,
        FETCH_EXCHANGE_RATES_FAILURE: (
            state,
            action: PayloadAction<string>
        ) => {
            state.error = action.payload;
        },
        SET_EXCHANGE_RATES: (state, action: PayloadAction<ExchangeRate[]>) => {
            Object.entries(action.payload).forEach(([code, currency]) => {
                const currencyToUpdate = state.currencies.find(
                    (item) => item.code === code
                );
                if (currencyToUpdate) {
                    currencyToUpdate.exchangeRate = parseFloat(
                        (currency as any).value.toFixed(2)
                    );
                }
            });
        },
        ADD_TO_FEATURED_CURRENCIES: (state, action) => {
            state.currencies = state.currencies.map((currency) => {
                if (currency.id === action.payload.id) {
                    return {
                        ...currency,
                        featured: true,
                    };
                }
                return currency;
            });
        },
        REMOVE_FROM_FEATURED_CURRENCIES: (state, action) => {
            state.currencies = state.currencies.map((currency) => {
                if (currency.id === action.payload.id) {
                    return {
                        ...currency,
                        featured: false,
                    };
                }
                return currency;
            });
        },
    },
});

function* fetchCurrenciesSaga() {
    try {
        const response: ExchangeRatesCurrenciesResponse = yield call(
            fetchExchangeRatesCurrencies
        );
        yield put(currenciesActions.SET_CURRENCIES(response));
    } catch (error: any) {
        yield put(currenciesActions.FETCH_CURRENCIES_FAILURE(error.message));
    }
}

export function* watchCurrencies() {
    yield takeLatest(currenciesActions.FETCH_CURRENCIES, fetchCurrenciesSaga);
}

function* fetchBaseCurrencySaga() {
    try {
        const response: string = yield call(fetchBaseCurrency);
        yield put(currenciesActions.SET_BASE_CURRENCY(response));
    } catch (error: any) {
        yield put(currenciesActions.FETCH_BASE_CURRENCY_FAILURE(error.message));
    }
}

export function* watchBaseCurrency() {
    yield takeLatest(
        currenciesActions.FETCH_BASE_CURRENCY,
        fetchBaseCurrencySaga
    );
}

function* fetchExchangeRatesSaga(
    action: PayloadAction<{
        baseCurrency: string;
        currencyCodes: string[];
    }>
) {
    const { baseCurrency, currencyCodes } = action.payload;
    try {
        const response: ExchangeRate[] = yield call(
            fetchExchangeRates,
            baseCurrency,
            currencyCodes
        );
        yield put(currenciesActions.SET_EXCHANGE_RATES(response));
    } catch (error: any) {
        yield put(
            currenciesActions.FETCH_EXCHANGE_RATES_FAILURE(error.message)
        );
    }
}

export function* watchExchangeRates() {
    yield takeLatest(
        currenciesActions.FETCH_EXCHANGE_RATES.type,
        fetchExchangeRatesSaga
    );
}

const { actions, reducer } = currenciesSlice;

export const currenciesActions = actions;
export const currenciesReducer = reducer;
