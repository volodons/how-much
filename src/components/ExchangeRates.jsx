import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Grid from '@mui/material/Grid';

import { currenciesActions } from '../redux/ducks/currenciesDuck';
import {
    StyledBox,
    StyledGrid,
    StyledTypographyHeading,
    StyledTypographyText,
    StyledButton,
    StyledStarIcon,
    StyledStarOutlineIcon,
} from '../styled/styledExchangeRates';

const ExchangeRates = () => {
    const dispatch = useDispatch();
    const allCurrencyCodes = useSelector(
        (state) => state.currencies.allCurrencyCodes
    );
    const allCurrencies = useSelector(
        (state) => state.currencies.allCurrencies
    );
    const baseCurrency = useSelector((state) => state.currencies.baseCurrency);

    const handleAddToFeaturedCurrencies = (currency) => {
        dispatch({
            type: currenciesActions.ADD_TO_FEATURED.type,
            payload: currency,
        });
    };

    const handleRemoveFromFeaturedCurrencies = (currency) => {
        dispatch({
            type: currenciesActions.REMOVE_FROM_FEATURED.type,
            payload: currency,
        });
    };

    useEffect(() => {
        dispatch({ type: currenciesActions.FETCH_ALL_CURRENCIES.type });
    }, []);

    useEffect(() => {
        dispatch({
            type: currenciesActions.FETCH_EXCHANGE_RATES.type,
            payload: { baseCurrency, allCurrencyCodes },
        });
    }, [baseCurrency]);

    return (
        <StyledBox>
            <StyledTypographyHeading variant="h2">
                Exchange Rates
            </StyledTypographyHeading>
            <Grid container>
                {baseCurrency &&
                    allCurrencies.map((currency) =>
                        currency.featured ? (
                            <StyledGrid container key={currency.id}>
                                <Grid item>
                                    <StyledButton
                                        onClick={() =>
                                            handleRemoveFromFeaturedCurrencies(
                                                currency
                                            )
                                        }
                                    >
                                        <StyledStarIcon />
                                    </StyledButton>
                                </Grid>
                                <Grid item>
                                    <StyledTypographyText>
                                        {currency.exchangeRate} {currency.code}{' '}
                                    </StyledTypographyText>
                                </Grid>
                            </StyledGrid>
                        ) : null
                    )}
                {baseCurrency &&
                    allCurrencies.map((currency) =>
                        !currency.featured ? (
                            <StyledGrid container key={currency.id}>
                                <Grid item>
                                    <StyledButton
                                        onClick={() =>
                                            handleAddToFeaturedCurrencies(
                                                currency
                                            )
                                        }
                                    >
                                        <StyledStarOutlineIcon />
                                    </StyledButton>
                                </Grid>
                                <Grid item>
                                    <StyledTypographyText>
                                        {currency.exchangeRate} {currency.code}{' '}
                                    </StyledTypographyText>
                                </Grid>
                            </StyledGrid>
                        ) : null
                    )}
            </Grid>
        </StyledBox>
    );
};

export default ExchangeRates;
