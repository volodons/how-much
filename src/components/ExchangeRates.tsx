import React from 'react';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Grid from '@mui/material/Grid';

import { RootState } from '../redux';
import { currenciesActions } from '../redux/ducks/currenciesDuck';
import {
    StyledBox,
    StyledGrid,
    StyledTypographyHeading,
    StyledTypographyText,
    StyledButton,
    StyledStarIcon,
    StyledStarOutlineIcon,
} from '../styled/styledExchangeRates';

interface Currency {
    id: number;
    code: string;
    exchangeRate: number;
    featured: boolean;
}

const ExchangeRates: React.FC = () => {
    const dispatch = useDispatch();
    const { currencyCodes, currencies, baseCurrency } = useSelector(
        (state: RootState) => state.currencies
    );

    const handleAddToFeaturedCurrencies = (currency: Currency) => {
        dispatch({
            type: currenciesActions.ADD_TO_FEATURED_CURRENCIES.type,
            payload: currency,
        });
    };

    const handleRemoveFromFeaturedCurrencies = (currency: Currency) => {
        dispatch({
            type: currenciesActions.REMOVE_FROM_FEATURED_CURRENCIES.type,
            payload: currency,
        });
    };

    useEffect(() => {
        dispatch({ type: currenciesActions.FETCH_CURRENCIES.type });
    }, []);

    useEffect(() => {
        dispatch({
            type: currenciesActions.FETCH_EXCHANGE_RATES.type,
            payload: { baseCurrency, currencyCodes },
        });
    }, [baseCurrency]);

    return (
        <StyledBox>
            <StyledTypographyHeading variant="h2">
                Exchange Rates
            </StyledTypographyHeading>
            <Grid container>
                {baseCurrency &&
                    currencies.map((currency: Currency) =>
                        currency.featured ? (
                            <StyledGrid container key={currency.id}>
                                <Grid item>
                                    <StyledButton
                                        onClick={() =>
                                            handleRemoveFromFeaturedCurrencies(
                                                currency
                                            )
                                        }
                                    >
                                        <StyledStarIcon />
                                    </StyledButton>
                                </Grid>
                                <Grid item>
                                    <StyledTypographyText>
                                        {currency.exchangeRate} {currency.code}{' '}
                                    </StyledTypographyText>
                                </Grid>
                            </StyledGrid>
                        ) : null
                    )}
                {baseCurrency &&
                    currencies.map((currency: Currency) =>
                        !currency.featured ? (
                            <StyledGrid container key={currency.id}>
                                <Grid item>
                                    <StyledButton
                                        onClick={() =>
                                            handleAddToFeaturedCurrencies(
                                                currency
                                            )
                                        }
                                    >
                                        <StyledStarOutlineIcon />
                                    </StyledButton>
                                </Grid>
                                <Grid item>
                                    <StyledTypographyText>
                                        {currency.exchangeRate} {currency.code}{' '}
                                    </StyledTypographyText>
                                </Grid>
                            </StyledGrid>
                        ) : null
                    )}
            </Grid>
        </StyledBox>
    );
};

export default ExchangeRates;
