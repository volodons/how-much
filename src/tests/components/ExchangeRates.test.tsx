import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { currenciesActions } from '../../redux/ducks/currenciesDuck';
import ExchangeRates from '../../components/ExchangeRates';

const createMockStore = configureMockStore();
const mockInitialState = {
    converter: {
        currencies: [],
        baseCurrency: { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
        baseCurrencyAmount: 0,
        targetCurrency: '',
        targetCurrencyAmount: 0,
        exchangeRate: 0,
        error: null,
    },
    currencies: {
        currencies: [
            { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
            { id: 2, code: 'USD', exchangeRate: 2, featured: false },
        ],
        currencyCodes: ['EUR', 'CAD'],
        baseCurrency: { id: 1, code: 'UAH', exchangeRate: 1, featured: true },
        error: null,
    },
};
const mockStore = createMockStore(mockInitialState);
const mockState = (mockStore as any).getState();

describe('ExchangeRates component', () => {
    it('should be a functional component', () => {
        expect(ExchangeRates).toBeInstanceOf(Function);
    });

    it('should render Exchange Rates heading', () => {
        render(
            <Provider store={mockStore}>
                <ExchangeRates />
            </Provider>
        );
        expect(screen.getByText('Exchange Rates')).toBeInTheDocument();
    });

    it('should add currency to favorites when star button is clicked', async () => {
        const mockDispatch = jest.fn();
        mockStore.dispatch = mockDispatch;

        render(
            <Provider store={mockStore}>
                <ExchangeRates />
            </Provider>
        );

        const starButton = screen.getByTestId('star-button');

        fireEvent.click(starButton);

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith({
                type: currenciesActions.ADD_TO_FEATURED_CURRENCIES.type,
                payload: {
                    id: 2,
                    code: 'USD',
                    exchangeRate: 2,
                    featured: false,
                },
            });
        });
    });

    it('should load currencies on mount', async () => {
        const mockDispatch = jest.fn();
        mockStore.dispatch = mockDispatch;

        render(
            <Provider store={mockStore}>
                <ExchangeRates />
            </Provider>
        );

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith({
                type: currenciesActions.FETCH_CURRENCIES.type,
            });
        });
    });

    it('should load exchange rates on mount', async () => {
        const mockDispatch = jest.fn();
        mockStore.dispatch = mockDispatch;

        const { baseCurrency, currencyCodes } = mockState.currencies;

        render(
            <Provider store={mockStore}>
                <ExchangeRates />
            </Provider>
        );

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith({
                type: currenciesActions.FETCH_EXCHANGE_RATES.type,
                payload: { baseCurrency, currencyCodes },
            });
        });
    });

    it('should remove currency from favorites when star button is clicked', async () => {
        const mockDispatch = jest.fn();
        mockStore.dispatch = mockDispatch;

        render(
            <Provider store={mockStore}>
                <ExchangeRates />
            </Provider>
        );

        const unstarButton = screen.getByTestId('unstar-button');

        fireEvent.click(unstarButton);

        await waitFor(() => {
            expect(mockDispatch).toHaveBeenCalledWith({
                type: currenciesActions.REMOVE_FROM_FEATURED_CURRENCIES.type,
                payload: {
                    id: 1,
                    code: 'UAH',
                    exchangeRate: 1,
                    featured: true,
                },
            });
        });
    });

    it('should match a snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={mockStore}>
                    <ExchangeRates />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
