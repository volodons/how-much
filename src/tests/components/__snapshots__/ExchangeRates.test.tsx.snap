// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeRates component should match a snapshot 1`] = `
<div
  className="MuiBox-root css-pw3vll"
  data-testid="exchange-rates-component"
>
  <h2
    className="MuiTypography-root MuiTypography-h2 css-1oqg8so-MuiTypography-root"
  >
    Exchange Rates
  </h2>
  <div
    className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-container css-4o2iwn-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-k9479y-MuiButtonBase-root-MuiButton-root"
          data-testid="unstar-button"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1h1jnqx-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-kazg0l-MuiTypography-root"
        >
          1
           
          UAH
           
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container css-4o2iwn-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-k9479y-MuiButtonBase-root-MuiButton-root"
          data-testid="star-button"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1h1jnqx-MuiSvgIcon-root"
            data-testid="StarOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28z"
            />
          </svg>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-kazg0l-MuiTypography-root"
        >
          2
           
          USD
           
        </p>
      </div>
    </div>
  </div>
</div>
`;
